blog/2022-11-03-short-circuiting.md

```diff
@@ -0,0 +1,156 @@
+---
+title: Array short-circuiting
+author: Philip Munksgaard
+description: In-place updates without overhead
+---
+
+# Situation
+
+Futhark is a parallel high-level array-oriented language.  Parallelism is explicitly
+expressed by the user in terms of the second-order array combinators such as
+`map`, `reduce` and `scan`.  Because parallelism is explicitly described, the
+compiler is able to make certain guarantees about the parallelism, such that the
+guarantee that there are no data races in the code.
+
+To see why this matters, consider this simple stencil function:
+
+```
+def stencil [n] (xs: *[n]i64): [n]i64 =
+  xs with [1:n-1] = map (\i -> xs[i-1] + xs[i] + xs[i+1]) (1..<n-1)
+```
+
+In a language without any data-race guarantees, or we were to implement this
+function as a parallel for loop in an imperative language, we would have
+data-dependency errors.  Specifically, it is possible for one thread to update
+an element of the array before another is able to read the original value.  This
+is also called an [anti-dependency or WAR (write after
+read)](https://en.wikipedia.org/wiki/Data_dependency#Anti-dependency).
+
+High-level languages such as Futhark are able to guarantee that this doesn't
+happen by transforming the code above into the following:
+
+```
+def stencil [n] (xs: *[n]i64): [n]i64 =
+  let tmp = map (\i -> xs[i-1] + xs[i] + xs[i+1]) (1..<n-1)
+  in xs with [1:n-1] = tmp
+```
+
+In other words, an auxilliary buffer is introduced and used to guarantee that
+the updates to `xs` do not interfere with the reads.
+
+# Problem
+
+The problem is that this intermediate array can be unnecessary.  Consider the
+this piece of code, which updates one row of a two-dimensional array:
+
+```
+def update [n][m] (i: i64) (xs: *[n][m]i64): [n][m]i64 =
+  xs with [i] = map (+1) xs[i]
+```
+
+As programmers, it is easy to determine that this update would be safe to do
+entirely in-place.  The updated values of the row each depend only on the
+previous value, and so they can be computed and written entirely in parallel
+with no risk of data-dependency errors.  Unfortunately, Futhark was not able to
+do so.  The conservative nature of high-level data-race guarantees, such as the
+ones Futhark provide, result in unnecessary memory being allocated and
+extraneous ceopies.
+
+This is further complicated by that fact that, in order to easily express
+complex computations like NW (TODO LINK?), Futhark supports flat slices, which
+make it hard for the compiler to verify that there are no data-dependencies.
+For instance, in our NW benchmark, the main loop boils down to a call like the
+following:
```

Comment:
I'd suggest handwaving the details as "more complex cases" (distinction between "slice" and "flat slice" is not something people care about).